src/QsCompiler/Transformations/ContentLifting.cs

```diff
@@ -456,12 +465,225 @@ public TypeTransformation(SyntaxTreeTransformation<TransformationState> parent)
+                                    .Zip(tupleType2.Item, (x, y) => this.ValidateAndCombineReturnTypes(x, y))
+                                    .ToImmutableArray()));
+                        }
+
+                        this.SharedState.IsValidScope = false;
+                        return type1;
+                    }
+                    else if (typeKind1 is ResolvedTypeKind.Operation opType1)
+                    {
+                        if (typeKind2 is ResolvedTypeKind.Operation opType2)
+                        {
+                            var subTypes = Tuple.Create(
+                                    this.ValidateAndCombineReturnTypes(opType1.Item1.Item1, opType2.Item1.Item1),
+                                    this.ValidateAndCombineReturnTypes(opType1.Item1.Item2, opType2.Item1.Item2));
+                            var commonInfo = CallableInformation.Common(new[] { opType1.Item2, opType2.Item2 });
+
+                            return ResolvedType.New(ResolvedTypeKind.NewOperation(subTypes, commonInfo));
+                        }
+
+                        this.SharedState.IsValidScope = false;
+                        return type1;
+                    }
+                    else if (typeKind1 is ResolvedTypeKind.Function funcType1)
+                    {
+                        if (typeKind2 is ResolvedTypeKind.Function funcType2)
+                        {
+                            return ResolvedType.New(ResolvedTypeKind.NewFunction(
+                                this.ValidateAndCombineReturnTypes(funcType1.Item1, funcType2.Item1),
+                                this.ValidateAndCombineReturnTypes(funcType1.Item2, funcType2.Item2)));
+                        }
+
+                        this.SharedState.IsValidScope = false;
+                        return type1;
+                    }
+                    else
+                    {
+                        if (!typeKind1.Equals(typeKind2))
+                        {
+                            this.SharedState.IsValidScope = false;
+                        }
+
+                        return type1;
+                    }
+                }
+
+                /// <inheritdoc/>
+                public override QsStatementKind OnReturnStatement(TypedExpression ex)
+                {
+                    if (!this.SharedState.IsReturnAllowed)
+                    {
+                        this.SharedState.IsValidScope = false;
+                    }
+                    else
+                    {
+                        if (this.SharedState.ReturnType == null)
+                        {
+                            this.SharedState.ReturnType = ex.ResolvedType;
+                        }
+                        else
+                        {
+                            this.SharedState.ReturnType = this.ValidateAndCombineReturnTypes(this.SharedState.ReturnType, ex.ResolvedType);
+                        }
+                    }
```

Comment:
Do we have any tests for lifting when return is allowed, or is that always disabled in our currently supported scenarios?

Reply:
Looks like we always pass false here... I'm curious when we anticipate we would start allowing returns from lifted functions, but in the meantime this is essentially untested dead code. Might be worth a comment calling it out as speculative implementation and highlighting when it might need to be used?

Reply:
There is not currently a way (and I can't think of any) to test the features of the base lifter directly. We can only test the features of the base lifter as they are used in derived lifters like that of the classical control lifter, which we do test. Since a lot of this PR was a refactoring of the base lifter to 1. clean it up and 2. prepare it for the derived lambda lifter, there were features added here that can't be tested directly until used by the lambda lifter. Return values are one and lifting in functions is another that I can think of.

Reply:
I've added a comment per your suggestion.